designs/replicated-pv/mayastor/pool-expansion.md

```diff
@@ -0,0 +1,157 @@
+
+4. A New Pool/DSP status field
+
+  - `max_growable_size`: This is the absolute maximum growable size. Even growth of additional 1 cluster would fail. Please note that this would not impact Active Application IO. It's just that the extended capacity is unusable.
+
+  - `disk_capacity` will show the size of the underlying device.
+
+  NOTE: `capacity` is the actual usable space of the Pool after reserving metadata from `disk_capacity`. So while growing user/admin should consider current `disk_capacity` and `max_growable_size`
+
+### Key Concepts
+
+**LVS Metadata restriction**:
+  When a mayastor pool is created, The metadata region is initialized which is immutable. Future expansion of the pool depends on the metadata reserved on create time. `max_growth` in the CR spec is used to reserve required metadata. All DiskPools created prior to this will have a limited grow capability as they would be not having enough metadata to accommodate bigger device size. The User/Admin might need to manually move replica off of the Pool and recreate the Pool by using `max_growth` if they want to grow older Pools.
+
+
+### Workflow
+
+1. **Diskpool CRD migration**:
+   - After the mayastor upgrade to release with this feature support. The existing CR will get a new field called `max_growable_size` in their status. Newer pool creation can have `max_growth` field in the spec to allow future growth.
+
+2. **Diskpool Creation**:
+    - The user/admin creates a diskpool yaml spec that contains the `max_growth` field indicating the expected growth in factor or absolute size.
+    - When the spec is applied, the diskpool operator picks up this request and dispatches the create pool request to the Mayastor agents that complete the diskpool provisioning.
+    - The spec for creating a pool with `max_growth` will look like below:
+
+```
+apiVersion: "openebs.io/v1beta3"
+kind: DiskPool
+metadata:
+  name: <pool-name>
+  namespace: <namespace>
+spec:
+  node: <node-name>
+  disks: ["/dev/disk/by-id/<id>"]
+  max_growth: "20x" or "6TiB"
+```
+   - The User/Admin can check `max_growable_size` of the newly provisioned pool. It should be honoring the `max_growth`.
+
+3. **Grow the DiskPool**:
+   - The User/Admin must expand the backing device (≤ `max_growable_size`).
+   - Then User/Admin can do either of the following to trigger mayastor pool grow workflow:
+     - **Annotate the DSP CR**
+       - add annotation `"openebs.io/grow-pool": true` on the CR to trigger reconciliation loop. The Operator removes the annotation when the grow succeeds or it sees error
+       which is deemed irrecoverable.
+     - **Use openebs plugin**
+       - Run kubectl openebs mayastor grow pool <pool-id> which will call grow pool api.
+
+
+## User Stories
+
+1. **Story 1**:
+  As a System Administrator, I notice few Pool's Allocation is nearing Capacity. It has higher committed size then capacity. Hence, I want to grow the pool to avoid thin provisioned replicas encountering ENOSPC.
```

Comment:
Is it a good idea to be able to check if pools are nearing their capacity using the plugin?

Reply:
Yes, users can do that. They can compare Allocated and Capacity They can also use monitoring stack.